src/Juvix/Core/IR/Subst.hs

```diff
@@ -0,0 +1,105 @@
+-- |
+-- - Runs a substitution algorithm over core
+module Juvix.Core.IR.Subst where
+
+import Control.Lens
+import qualified Data.HashSet as Set
+import qualified Juvix.Core.Common.NameSymbol as NameSymbol
+import qualified Juvix.Core.IR.Types as Types
+import qualified Juvix.Core.Usage as Usage
+import Juvix.Library
+import qualified Juvix.Library.HashMap as Map
+
+-- eventually a lot of this code will be replaced by a proper
+-- inliner strategy, but this code will serve as the basis for
+-- said iterative design
+
+data Key
+  = Bound Natural
+  | Global NameSymbol.T
+  deriving (Show, Eq, Generic)
+
+instance Hashable Key
+
+data T primTy primVal
+  = T
+      { _depth :: Natural,
+        _sub :: Map.T Key (Types.Elim primTy primVal)
+      }
+  deriving (Show)
+
+makeLenses ''T
+
+-- useless with debrunj indincies ☹
+type InScopeSet = Set.HashSet Types.Name
+
+-- TODO ∷
+-- - add a context for this to run with
+-- - Turn the Context, T, InScopeSet to being in the Env
+-- - when seeing a new binding determine if we should add that to the Τ type
+--   + For now we just ignore extra values to inline, and continue on our way
+f ::
+  T primTy primVal -> Types.Term primTy primVal -> Types.Term primTy primVal
+f _ (Types.PrimTy ty) =
+  Types.PrimTy ty
+f _ (Types.Star uni) =
+  Types.Star uni
+f subst (Types.Pi usage typ body) =
+  Types.Pi usage (f subst typ) (f (over depth succ subst) body)
+f subst (Types.Lam body) =
+  Types.Lam (f (over depth succ subst) body)
+f subst (Types.Let usage bound body)
+  | inlineP usage bound =
```

Comment:
honestly I find it odd that this function makes inlining decisions, won't that result in a fair bit of duplicate work (checking `inlineP` at least) if we substitute several times over the same term?

Reply:
Ideally we'll have annotation extension to this, should it should be better. The GHC inliner does something very similar, as it checks things as it goes.